lib/IRGen/GenMeta.cpp

```diff
@@ -1532,27 +1532,204 @@ void IRGenModule::addFieldTypes(ArrayRef<CanType> fieldTypes) {
+  IGF.Builder.CreateLifetimeStart(fields,
+                  IGF.IGM.getPointerSize() * storedProperties.size());
+  fields = IGF.Builder.CreateStructGEP(fields, 0, Size(0));
+
+  unsigned index = 0;
+  for (auto prop : storedProperties) {
+    auto propTy = T.getFieldType(prop, IGF.getSILModule());
+    llvm::Value *metadata = emitTypeLayoutRef(IGF, propTy, collector);
+    Address field = IGF.Builder.CreateConstArrayGEP(fields, index,
+                                                    IGF.IGM.getPointerSize());
+    IGF.Builder.CreateStore(metadata, field);
+    ++index;
+  }
+
+  // Ask the runtime to lay out the struct or class.
+  auto numFields = IGF.IGM.getSize(Size(storedProperties.size()));
+
+  if (auto *classDecl = dyn_cast<ClassDecl>(target)) {
+    // Compute class layout flags.
+    ClassLayoutFlags flags = ClassLayoutFlags::Swift5Algorithm;
+    if (!doesClassMetadataRequireRelocation(IGF.IGM, classDecl))
+      flags |= ClassLayoutFlags::HasStaticVTable;
+
+    // Get the superclass metadata, if the class has one.
+    llvm::Value *superclassMetadata;
+    if (auto superclassType = classDecl->getSuperclass()) {
+      superclassType = classDecl->mapTypeIntoContext(superclassType);
+
+      auto request = DynamicMetadataRequest::getNonBlocking(
+                               MetadataState::NonTransitiveComplete, collector);
+
+      superclassMetadata =
+        emitClassHeapMetadataRef(IGF, superclassType->getCanonicalType(),
+                                 MetadataValueType::TypeMetadata,
+                                 request,
+                                 /*allowUninit*/ false);
+    } else {
+      superclassMetadata =
+        llvm::ConstantPointerNull::get(IGF.IGM.TypeMetadataPtrTy);
+    }
+
+    // Call swift_initClassMetadata().
+    IGF.Builder.CreateCall(IGF.IGM.getInitClassMetadataFn(),
+                           {metadata, superclassMetadata,
+                            IGF.IGM.getSize(Size(uintptr_t(flags))),
+                            numFields, fields.getAddress(), fieldVector});
+  } else {
+    assert(isa<StructDecl>(target));
+
+    // Compute struct layout flags.
+    StructLayoutFlags flags = StructLayoutFlags::Swift5Algorithm;
+    if (isVWTMutable)
+      flags |= StructLayoutFlags::IsVWTMutable;
+
+    // Call swift_initStructMetadata().
+    IGF.Builder.CreateCall(IGF.IGM.getInitStructMetadataFn(),
+                           {metadata, IGF.IGM.getSize(Size(uintptr_t(flags))),
+                            numFields, fields.getAddress(), fieldVector});
+  }
+
+  IGF.Builder.CreateLifetimeEnd(fields,
+                  IGF.IGM.getPointerSize() * storedProperties.size());
+}
+
+static void emitInitializeMetadata(IRGenFunction &IGF,
+                                   NominalTypeDecl *nominalDecl,
+                                   llvm::Value *metadata,
+                                   bool isVWTMutable,
+                                   MetadataDependencyCollector *collector) {
```

Comment:
Should this be called `emitInitializeValueMetadata` to make it clear that it's not for classes?

Reply:
Good idea